### PR TITLE
Fixes #1763 - Flaky test: SimpleTimerTest

### LINKED ISSUE #1763: Flaky test: SimpleTimerTest
**Test name(s)**
 - org.apache.accumulo.server.util.time.SimpleTimerTest.testFailure

**Describe the failure observed**
Stacktrace
```java
java.lang.AssertionError
	at org.apache.accumulo.server.util.time.SimpleTimerTest.testFailure(SimpleTimerTest.java:98)
```
Standard Output
```
2020-10-29T03:28:54,071 [time.SimpleTimer] WARN : Asked to create SimpleTimer with thread pool size 2, existing instance has 1
```

**Testing Environment:**

 - ASF Jenkins
 - Commit ea5fa7d1da90da9dfe1128c9561022d310102777
```
Executing Maven:  -B -f /home/jenkins/workspace/Accumulo/main/pom.xml -V -e -ntp clean verify -Dstyle.color=always -Dfailsafe.rerunFailingTestsCount=5
Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
Maven home: /usr/local/asfpackages/maven/apache-maven-3.6.3
Java version: 11.0.6, vendor: Oracle Corporation, runtime: /usr/local/asfpackages/java/jdk-11.0.6
Default locale: en_US, platform encoding: ISO-8859-1
OS name: "linux", version: "4.15.0-74-generic", arch: "amd64", family: "unix"
```

**What have you tried already?**
I re-ran the test and it usually passes.

**Additional context**
There's a good chance this is just a flaky test, because of resource contention on the server. The warning seems related to the `testSingleton` test case, and not the `testFailure` test case that failed.

One thing that might be possible to do to is to wrap the assert statement in a loop, with a delay of `PAD`, until it passes, and then add a JUnit timeout to the test of some reasonable length (5 seconds?). That way, if the pass condition isn't met immediately, the test won't fail, as long as it is met within the timeout period.

### PR DESCRIPTION
**Pull request by DomGarguilo**
Did my best to resolve #1763. Replaced the assert statements with an if statement so that they could be repeated if they don't pass immediately. The test will repeat until passed or until the added timeout occurs. As suggested, I left the assert statement after the loop which verifies the thread pool size remains unchanged after the previous operations. Not sure if I should have kept the other asserts as well but I don't see the point if the same conditions were verified already. Please suggest any changes or improvements.

### CODE REVIEW

File: server/base/src/test/java/org/apache/accumulo/server/util/time/SimpleTimerTest.java [hunk 1]
```diff
@@ -98,11 +97,15 @@ public void testFailure() throws InterruptedException {
...
-    assertSame(t, t2);
+  @Test(timeout = 5000)
+  public void testSingleton() throws InterruptedException {
+    while (true) {
+      SimpleTimer t2 = SimpleTimer.getInstance(2);
+      if((SimpleTimer.getInstanceThreadPoolSize() == 1) && (t == t2)) {
+        break;
+      }
+      Thread.sleep(PAD);
+    }
```
Comment (ctubbsii) [inline server/base/src/test/java/org/apache/accumulo/server/util/time/SimpleTimerTest.java:108]: This is the right strategy (albeit with a few extra parenthesis than necessary on that `if` statement :wink: )

**Review by ctubbsii — CHANGES_REQUESTED** [review]
Hi @DomGarguilo . Thanks for the PR! You have the correct strategy in place. However, the test case that you modified was probably the only test case for this class that isn't prone to the type of failure that was reported. It should pass every time. If you look at the other 3 tests, you can see that they all employ a "sleep" already, in order to try to catch the test in the correct state. However, since that's flaky (it may take slightly longer for the task to start), those other 3 tests are the ones that should employ the method you applied here (instead of this one).

EDIT: also, please make sure the changed files are formatted to our coding standards by running `mvn package -DskipTests` at the root of the project, and including any changes made by our formatting plugins to your commits / PR.

File: server/base/src/test/java/org/apache/accumulo/server/util/time/SimpleTimerTest.java (hunk 1 above)
Comment (DomGarguilo) [reply server/base/src/test/java/org/apache/accumulo/server/util/time/SimpleTimerTest.java:108]: Oops, I see I was off-base with this one. Will do my best to correct these issues with your suggestions in mind :+1:

**Comment by DomGarguilo** [comment]
Oops, I see I was off-base with this one. Will do my best to correct these issues with your suggestions in mind +1

**Comment by DomGarguilo** [comment]
I implemented the fix in the correct test cases. There were a few things that I was unsure about though. Not sure if I should extract the code containing the while(true) and create a separate method in order to reuse that portion of code. Not sure if I should add a local variable for the timeout length (even though there is one test case that requires a longer timeout). I added a timeout to testSingleton() which probably doesn't need it so I am unsure whether I should remove it or keep it in.

File: server/base/src/test/java/org/apache/accumulo/server/util/time/SimpleTimerTest.java [hunk 2]
```diff
@@ -70,35 +69,50 @@ public void run() {
...
     Incrementer r = new Incrementer(i);
     t.schedule(r, DELAY);
     Thread.sleep(DELAY + PAD);
-    assertEquals(1, i.get());
+    while (true) {
+      if (i.get() == 1) {
+        break;
+      }
+      Thread.sleep(PAD);
+    }
```
Comment (ctubbsii) [inline server/base/src/test/java/org/apache/accumulo/server/util/time/SimpleTimerTest.java:83]: Just realized that you can simplify this to `while (i.get() != 1) { sleep }`; you can do similar things for the other tests as well.

**Review by ctubbsii — CHANGES_REQUESTED** [review]

File: server/base/src/test/java/org/apache/accumulo/server/util/time/SimpleTimerTest.java (hunk 2 above)
Comment (DomGarguilo) [reply server/base/src/test/java/org/apache/accumulo/server/util/time/SimpleTimerTest.java:83]: Was considering that as well. I will add this change.

**Comment by ctubbsii** [comment]
> Not sure if I should extract the code containing the while(true) and create a separate method in order to reuse that portion of code.

No. These changes aren't big enough to warrant that.

> Not sure if I should add a local variable for the timeout length (even though there is one test case that requires a longer timeout).

No. Your timeouts are fine.

> I added a timeout to testSingleton() which probably doesn't need it so I am unsure whether I should remove it or keep it in.

In general, we try to make minimal unrelated changes, to make it easier to review, but it's often a judgement call. In this case, I would leave it in. It's a tiny change, and is related to your adding a timeout to all the other test cases.

**Review by ctubbsii — APPROVED** [review]

**Comment by ctubbsii** [comment]
Thanks for improving the test, @DomGarguilo !

**Pull request merged**